Segmentation fault  mailparse 3.1.9

Hi, I’m the maintainer of [php-mime-mail-parser](https://github.com/php-mime-mail-parser/php-mime-mail-parser), which depends on mailparse. When running our full test suite under PHP 8.5 with mailparse 3.1.9, all tests pass but PHP crashes immediately after PHPUnit finishes.

**Details**:
- PHP 8.0–8.4 work fine with earlier mailparse versions.
- The crash only occurs on PHP 8.5 with mailparse 3.1.9.
- PHPUnit 9.6.29 is used.

Here is the issue:
https://github.com/php-mime-mail-parser/php-mime-mail-parser/actions/runs/19296608895/job/55180068007?pr=473

```
PHPUnit 9.6.22 by Sebastian Bergmann and contributors.

...............................................................  63 / 159 ( 39%)
............................................................... 126 / 159 ( 79%)
.................................                               159 / 159 (100%)

Time: 00:00.516, Memory: 31.12 MB

OK (159 tests, 1915 assertions)
/home/runner/work/_temp/6157c364-f825-41de-9817-34a9b5a077e6.sh: line 1:  5366 Segmentation fault      (core dumped) vendor/bin/phpunit
Error: Process completed with exit code 139.
```

Thanks for looking into this — I’m happy to provide more details or help debug if needed.

--- a/.gitattributes
+++ b/.gitattributes
@@ -1,0 +1,4 @@
+# Keep test files and test data with LF line endings on all platforms
+# to ensure byte offsets and string lengths match expected values.
+*.phpt text eol=lf
+tests/testdata/* -text

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -46,7 +46,7 @@ jobs:
         uses: actions/checkout@v2
       - name: Setup PHP
         id: setup-php
-        uses: cmb69/setup-php-sdk@v0.2
+        uses: php/setup-php-sdk@c391923f79202e73f64b8cb15844fee60cbc6c94
         with:
           version: ${{matrix.version}}
           arch: ${{matrix.arch}}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -62,4 +62,6 @@ jobs:
       - name: make
         run: nmake
       - name: test
-        run: nmake test TESTS="-d extension=${{steps.setup-php.outputs.prefix}}\ext\php_mbstring.dll --show-diff tests"
+        run: |
+          set PATH=%PATH%;${{steps.setup-php.outputs.prefix}}\ext
+          nmake test TESTS="-d extension=${{steps.setup-php.outputs.prefix}}\ext\php_mbstring.dll --show-diff tests"

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -4,13 +4,13 @@ jobs:
   ubuntu:
     strategy:
       matrix:
-          version: ['7.3', '7.4', '8.0', '8.1', '8.2', '8.3']
+          version: ['7.3', '7.4', '8.0', '8.1', '8.2', '8.3', '8.4', '8.5']
     runs-on: ubuntu-latest
     steps:
       - name: Install re2c
         run: sudo apt-get install -y re2c
       - name: Checkout mailparse
-        uses: actions/checkout@v2
+        uses: actions/checkout@v4
       - name: Setup PHP
         uses: shivammathur/setup-php@v2
         with:
@@ -34,19 +34,18 @@ jobs:
       run:
         shell: cmd
     strategy:
+      fail-fast: false
       matrix:
-          version: ["7.4", "8.0"]
+          version: ["8.1", "8.2", "8.3", "8.4", "8.5"]
           arch: [x64, x86]
           ts: [ts]
     runs-on: windows-latest
     steps:
-      - name: Configure Git
-        run: git config --global core.autocrlf false
       - name: Checkout mailparse
-        uses: actions/checkout@v2
+        uses: actions/checkout@v4
       - name: Setup PHP
         id: setup-php
-        uses: php/setup-php-sdk@c391923f79202e73f64b8cb15844fee60cbc6c94
+        uses: php/setup-php-sdk@fix/invoke-webrequest-tls
         with:
           version: ${{matrix.version}}
           arch: ${{matrix.arch}}

--- a/mailparse.c
+++ b/mailparse.c
@@ -122,7 +122,10 @@ ZEND_RSRC_DTOR_FUNC(mimepart_dtor)
 {
 	php_mimepart *part = res->ptr;
 
-	php_mimepart_free(part);
+	if (part != NULL) {
+		res->ptr = NULL;
+		php_mimepart_free(part);
+	}
 }
 
 PHP_INI_BEGIN()

--- a/php_mailparse_mime.c
+++ b/php_mailparse_mime.c
@@ -321,8 +321,14 @@ PHP_MAILPARSE_API void php_mimepart_free(php_mimepart *part)
 	zval *childpart_z;
 	HashPosition pos;
 
-	/* Recursively free children, NULLing their resource pointers
-	 * to prevent double-free from the resource list cleanup */
+	/* Prevent the resource destructor from freeing this part again */
+	if (part->rsrc && part->rsrc->ptr == part) {
+		part->rsrc->ptr = NULL;
+	}
+
+	/* Recursively free children, NULLing their resource pointers to prevent
+	 * double-free from the resource list cleanup, and releasing the resource
+	 * list entry so the refcount doesn't leak */
 	zend_hash_internal_pointer_reset_ex(&part->children, &pos);
 	while ((childpart_z = zend_hash_get_current_data_ex(&part->children, &pos)) != NULL) {
 		zend_resource *child_res = Z_RES_P(childpart_z);
@@ -331,6 +337,7 @@ PHP_MAILPARSE_API void php_mimepart_free(php_mimepart *part)
 			child_res->ptr = NULL;
 			php_mimepart_free(child);
 		}
+		zend_list_delete(child_res);
 		zend_hash_move_forward_ex(&part->children, &pos);
 	}
 

--- a/php_mailparse_mime.c
+++ b/php_mailparse_mime.c
@@ -326,18 +326,14 @@ PHP_MAILPARSE_API void php_mimepart_free(php_mimepart *part)
 		part->rsrc->ptr = NULL;
 	}
 
-	/* Recursively free children, NULLing their resource pointers to prevent
-	 * double-free from the resource list cleanup, and releasing the resource
-	 * list entry so the refcount doesn't leak */
+	/* Release child resources via zval_ptr_dtor, which respects refcounts:
+	 * children with no external references are freed immediately, while
+	 * children held by userland (e.g. via mailparse_msg_get_part) stay
+	 * alive until their own refcount reaches zero. The idempotent
+	 * mimepart_dtor (res->ptr NULL guard) prevents double-free at shutdown. */
 	zend_hash_internal_pointer_reset_ex(&part->children, &pos);
 	while ((childpart_z = zend_hash_get_current_data_ex(&part->children, &pos)) != NULL) {
-		zend_resource *child_res = Z_RES_P(childpart_z);
-		php_mimepart *child = (php_mimepart *)child_res->ptr;
-		if (child != NULL) {
-			child_res->ptr = NULL;
-			php_mimepart_free(child);
-		}
-		zend_list_delete(child_res);
+		zval_ptr_dtor(childpart_z);
 		zend_hash_move_forward_ex(&part->children, &pos);
 	}
 

--- a/php_mailparse_mime.c
+++ b/php_mailparse_mime.c
@@ -321,10 +321,16 @@ PHP_MAILPARSE_API void php_mimepart_free(php_mimepart *part)
 	zval *childpart_z;
 	HashPosition pos;
 
-	/* free contained parts */
+	/* Recursively free children, NULLing their resource pointers
+	 * to prevent double-free from the resource list cleanup */
 	zend_hash_internal_pointer_reset_ex(&part->children, &pos);
 	while ((childpart_z = zend_hash_get_current_data_ex(&part->children, &pos)) != NULL) {
-		zval_ptr_dtor(childpart_z);
+		zend_resource *child_res = Z_RES_P(childpart_z);
+		php_mimepart *child = (php_mimepart *)child_res->ptr;
+		if (child != NULL) {
+			child_res->ptr = NULL;
+			php_mimepart_free(child);
+		}
 		zend_hash_move_forward_ex(&part->children, &pos);
 	}
 

--- a/tests/gh44.phpt
+++ b/tests/gh44.phpt
@@ -1,0 +1,59 @@
+--TEST--
+GH issue #44 (Segmentation fault in mimepart resource destructor during shutdown)
+--SKIPIF--
+<?php
+if (!extension_loaded("mailparse")) die("skip mailparse extension not available");
+?>
+--FILE--
+<?php
+$mime = "Content-Type: multipart/mixed; boundary=\"outer\"\r\n" .
+    "\r\n" .
+    "--outer\r\n" .
+    "Content-Type: multipart/alternative; boundary=\"inner\"\r\n" .
+    "\r\n" .
+    "--inner\r\n" .
+    "Content-Type: text/plain\r\n" .
+    "\r\n" .
+    "Hello plain\r\n" .
+    "--inner\r\n" .
+    "Content-Type: text/html\r\n" .
+    "\r\n" .
+    "<b>Hello html</b>\r\n" .
+    "--inner--\r\n" .
+    "--outer\r\n" .
+    "Content-Type: application/octet-stream\r\n" .
+    "Content-Transfer-Encoding: base64\r\n" .
+    "\r\n" .
+    "SGVsbG8=\r\n" .
+    "--outer--\r\n";
+
+/* Test 1: Parse multipart message, let resource cleanup happen at shutdown */
+$msg1 = mailparse_msg_create();
+mailparse_msg_parse($msg1, $mime);
+$struct = mailparse_msg_get_structure($msg1);
+echo "structure: " . implode(", ", $struct) . "\n";
+
+/* Test 2: Parse and explicitly free */
+$msg2 = mailparse_msg_create();
+mailparse_msg_parse($msg2, $mime);
+mailparse_msg_free($msg2);
+
+/* Test 3: Get child parts (adds refcount), then let shutdown clean up */
+$msg3 = mailparse_msg_create();
+mailparse_msg_parse($msg3, $mime);
+$parts = [];
+foreach (mailparse_msg_get_structure($msg3) as $part_id) {
+    $parts[] = mailparse_msg_get_part($msg3, $part_id);
+}
+
+/* Test 4: Multiple messages with interleaved resource handles */
+for ($i = 0; $i < 5; $i++) {
+    $m = mailparse_msg_create();
+    mailparse_msg_parse($m, $mime);
+}
+
+echo "ok\n";
+?>
+--EXPECT--
+structure: 1, 1.1, 1.1.1, 1.1.2, 1.2
+ok

--- a/tests/gh44.phpt
+++ b/tests/gh44.phpt
@@ -6,54 +6,51 @@ if (!extension_loaded("mailparse")) die("skip mailparse extension not available"
 ?>
 --FILE--
 <?php
-$mime = "Content-Type: multipart/mixed; boundary=\"outer\"\r\n" .
-    "\r\n" .
-    "--outer\r\n" .
-    "Content-Type: multipart/alternative; boundary=\"inner\"\r\n" .
-    "\r\n" .
-    "--inner\r\n" .
-    "Content-Type: text/plain\r\n" .
-    "\r\n" .
-    "Hello plain\r\n" .
-    "--inner\r\n" .
-    "Content-Type: text/html\r\n" .
-    "\r\n" .
-    "<b>Hello html</b>\r\n" .
-    "--inner--\r\n" .
-    "--outer\r\n" .
-    "Content-Type: application/octet-stream\r\n" .
-    "Content-Transfer-Encoding: base64\r\n" .
-    "\r\n" .
-    "SGVsbG8=\r\n" .
-    "--outer--\r\n";
+/* Generate a multipart message with >300 parts to trigger MAXPARTS.
+ * This exercises the mailparse_msg_parse_file error path which calls
+ * php_mimepart_free() directly, leaving child resources in the resource
+ * list. Without the fix, these dangling resources cause a use-after-free
+ * during shutdown in zend_close_rsrc_list. */
 
-/* Test 1: Parse multipart message, let resource cleanup happen at shutdown */
-$msg1 = mailparse_msg_create();
-mailparse_msg_parse($msg1, $mime);
-$struct = mailparse_msg_get_structure($msg1);
+$boundary = "test_boundary";
+$mime = "Content-Type: multipart/mixed; boundary=\"$boundary\"\r\n\r\n";
+for ($i = 0; $i < 302; $i++) {
+    $mime .= "--$boundary\r\nContent-Type: text/plain\r\n\r\npart $i\r\n";
+}
+$mime .= "--$boundary--\r\n";
+
+$tmpfile = tempnam(sys_get_temp_dir(), 'mp_gh44_');
+file_put_contents($tmpfile, $mime);
+
+/* Parse the oversized message - triggers "MIME message too complex" warning
+ * and the error path calls php_mimepart_free(part) directly, freeing the
+ * mimepart struct but leaving child zend_resource entries in the list */
+$result = @mailparse_msg_parse_file($tmpfile);
+echo "parse_file returned: " . var_export($result, true) . "\n";
+
+/* Allocate many messages to reuse freed memory from the failed parse above.
+ * This increases the chance that the dangling resource pointers from the
+ * failed parse will reference reused/corrupted memory at shutdown. */
+$msgs = [];
+for ($i = 0; $i < 50; $i++) {
+    $m = mailparse_msg_create();
+    mailparse_msg_parse($m, "Content-Type: multipart/mixed; boundary=\"b\"\r\n\r\n" .
+        "--b\r\nContent-Type: text/plain\r\n\r\nhello\r\n" .
+        "--b\r\nContent-Type: text/html\r\n\r\n<b>hi</b>\r\n--b--\r\n");
+    $msgs[] = $m;
+}
+
+/* Also test normal multipart parsing and cleanup */
+$msg = mailparse_msg_create();
+mailparse_msg_parse($msg, "Content-Type: multipart/mixed; boundary=\"x\"\r\n\r\n" .
+    "--x\r\nContent-Type: text/plain\r\n\r\nhello\r\n--x--\r\n");
+$struct = mailparse_msg_get_structure($msg);
 echo "structure: " . implode(", ", $struct) . "\n";
 
-/* Test 2: Parse and explicitly free */
-$msg2 = mailparse_msg_create();
-mailparse_msg_parse($msg2, $mime);
-mailparse_msg_free($msg2);
-
-/* Test 3: Get child parts (adds refcount), then let shutdown clean up */
-$msg3 = mailparse_msg_create();
-mailparse_msg_parse($msg3, $mime);
-$parts = [];
-foreach (mailparse_msg_get_structure($msg3) as $part_id) {
-    $parts[] = mailparse_msg_get_part($msg3, $part_id);
-}
-
-/* Test 4: Multiple messages with interleaved resource handles */
-for ($i = 0; $i < 5; $i++) {
-    $m = mailparse_msg_create();
-    mailparse_msg_parse($m, $mime);
-}
-
+unlink($tmpfile);
 echo "ok\n";
 ?>
 --EXPECT--
-structure: 1, 1.1, 1.1.1, 1.1.2, 1.2
+parse_file returned: false
+structure: 1, 1.1
 ok

